Npapi based daemon optimize

stephenxs opened the pull request:
I would like to merge these 3 commits into the [[Mellanox] optimize new platform api #3289](https://github.com/Azure/sonic-buildimage/pull/3289/files)
1. refactor reboot-cause handling since too many reset cause files on variant SKUs to use a series of coalescing if-else to handle.
2. optimize sfp initialization and reading.

Commit:
[sonic_platform]only initialize sfp modules when related apis are called

Commit:
[sonic_platform]chassis: refractor reboot cause handling. using a dictionary to handle reboot cause

Commit:
[sonic_platform] sfp: optimizing eeprom data read by combining multiple call to ethtool into one

keboliu commented on platform/mellanox/mlnx-platform-api/sonic_platform/chassis.py:
```diff
@@ -148,6 +143,58 @@ def initialize_components_list(self):
+        Returns:
+            An integer, the number of sfps available on this chassis
+        """
+        if self.sfp_module is None:
```
suggest adding a dedicated flag for sfp init, instead of using this self.sfp_module.

stephenxs replied:
fixed.  an explicit flag which represents whether SFP module initialized has been introduced.

keboliu commented on platform/mellanox/mlnx-platform-api/sonic_platform/sfp.py:
```diff
@@ -63,6 +63,24 @@
+# in function get_transceiver_info and get_transceiver_bulk_status
+# XCVR_INTERFACE_DATA_SIZE stands for the max size to be read
+# this variable is only used by get_transceiver_info.
+# be attention that each time some new value added to the function 
```
change "be attention" to "please be noted" ?

stephenxs replied:
fixed.

Commit:
[sonic_platform]introduce an explicit flag indicating whether sfp has been initialized

Commit:
[chassis]
don't initialize sfp_event until it is called
deinitialize sfp_event when chassis destructed so that sdk resources can be correctly released.
